### PR TITLE
Move the common code from testcases to the fixtures.

### PR DESCRIPTION
**Pull request by rraghav-cisco**
### Description of PR

Summary:
Snappi reboot testcases don't save the config before rebooting the DUT. This causes the DUT to loose all the config, and the test fails with "ARP is not resolved" errors.

This PR addresses this issue by saving the config before any reboot. Also this PR moves the code that is reused in multiple tests to a common code.

### Type of change

- [X] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [X] 202405

### CODE REVIEW

File: tests/common/snappi_tests/snappi_fixtures.py [hunk 1]
```diff
@@ -19,6 +19,12 @@
 from tests.snappi_tests.variables import dut_ip_start, snappi_ip_start, prefix_length, \
     dut_ipv6_start, snappi_ipv6_start, v6_prefix_length, pfcQueueGroupSize, \
     pfcQueueValueDict          # noqa: F401
+from tests.snappi_tests.variables import MULTIDUT_PORT_INFO, MULTIDUT_TESTBED
+from tests.snappi_tests.files.helper import skip_warm_reboot
```
Comment (yutongzhang-microsoft) [inline tests/common/snappi_tests/snappi_fixtures.py:23]: We do not permit cross-module dependencies within the repo. This means that in the `common` folder, functions from feature folders like `snappi_tests` should not be imported. Therefore, please relocate the shared functions in `snappi_tests` to a common location.

**Review by yutongzhang-microsoft — CHANGES_REQUESTED** [review]

File: tests/common/snappi_tests/snappi_fixtures.py (hunk 1 above)
Comment (rraghav-cisco) [reply tests/common/snappi_tests/snappi_fixtures.py:23]: Working on the move to other file. I will be moving the new fixture functions to tests/snappi_tests/files/helper.py. Pls let me know if that is ok. thanks.

Comment (yutongzhang-microsoft) [reply tests/common/snappi_tests/snappi_fixtures.py:23]: Sure, please go ahead.

**Review by yutongzhang-microsoft — APPROVED** [review]

File: tests/common/snappi_tests/snappi_fixtures.py (hunk 1 above)
Comment (rraghav-cisco) [reply tests/common/snappi_tests/snappi_fixtures.py]: @yutongzhang-microsoft : Completed the move. Thanks. The failures below are not related to this change, they are due to unconfigured-RP at that time.

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================================================================================================== PASSES ===========================================================================================================================
___________________________________________________________________________________________ test_pfc_pause_single_lossless_prio[multidut_port_info1-yy39top-lc4|3] ___________________________________________________________________________________________
___________________________________________________________________________________________ test_pfc_pause_single_lossless_prio[multidut_port_info1-yy39top-lc4|4] ___________________________________________________________________________________________
__________________________________________________________________________________________________ test_pfc_pause_multi_lossless_prio[multidut_port_info1] ___________________________________________________________________________________________________
_____________________________________________________________________________________ test_pfc_pause_single_lossless_prio_reboot[multidut_port_info1-cold-yy39top-lc4|3] _____________________________________________________________________________________
_____________________________________________________________________________________ test_pfc_pause_single_lossless_prio_reboot[multidut_port_info1-cold-yy39top-lc4|4] _____________________________________________________________________________________
____________________________________________________________________________________________ test_pfc_pause_multi_lossless_prio_reboot[multidut_port_info1-cold] _____________________________________________________________________________________________
-------------------------------------------------------------------------------- generated xml file: /run_logs/ixia/pfcwd-m2o/2024-10-29-20-07-30/tr_2024-10-29-20-07-30.xml ---------------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
23:18:54 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio[multidut_port_info1-yy39top-lc4|3]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio[multidut_port_info1-yy39top-lc4|4]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_multi_lossless_prio[multidut_port_info1]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio_reboot[multidut_port_info1-cold-yy39top-lc4|3]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio_reboot[multidut_port_info1-cold-yy39top-lc4|4]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_multi_lossless_prio_reboot[multidut_port_info1-cold]
SKIPPED [4] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py:138: Reboot type warm is not supported on cisco-8000 switches
SKIPPED [4] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py:138: Reboot type fast is not supported on cisco-8000 switches
SKIPPED [2] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py:197: Reboot type warm is not supported on cisco-8000 switches
SKIPPED [2] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py:197: Reboot type fast is not supported on cisco-8000 switches
ERROR snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio[multidut_port_info0-yy39top-lc4|3] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost yy39top-lc4>']" failed with exit code "1"
ERROR snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio[multidut_port_info0-yy39top-lc4|4] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost yy39top-lc4>']" failed with exit code "1"
ERROR snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio[multidut_port_info1-yy39top-lc4|3] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost yy39top-lc4>']" failed with exit code "1"
ERROR snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_multi_lossless_prio[multidut_port_info1] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost yy39top-lc4>']" failed with exit code "1"
ERROR snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_multi_lossless_prio_reboot[multidut_port_info1-fast] - Failed: Not all ports that are admin up on are operationally up
FAILED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio[multidut_port_info0-yy39top-lc4|3] - Failed: PFC pause frames should be transmitted and counted in TX PFC counters for priority 3
FAILED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio[multidut_port_info0-yy39top-lc4|4] - Failed: PFC pause frames should be transmitted and counted in TX PFC counters for priority 4
FAILED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_multi_lossless_prio[multidut_port_info0] - Failed: PFC pause frames should be transmitted and counted in TX PFC counters for priority 3
FAILED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio_reboot[multidut_port_info0-cold-yy39top-lc4|3] - Failed: PFC pause frames should be transmitted and counted in TX PFC counters for priority 3
FAILED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_single_lossless_prio_reboot[multidut_port_info0-cold-yy39top-lc4|4] - Failed: PFC pause frames should be transmitted and counted in TX PFC counters for priority 4
FAILED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py::test_pfc_pause_multi_lossless_prio_reboot[multidut_port_info0-cold] - Failed: PFC pause frames should be transmitted and counted in TX PFC counters for priority 3
======================================================================================== 6 failed, 6 passed, 12 skipped, 16 warnings, 5 errors in 11482.18s (3:11:22) ========================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$

```

Comment (rraghav-cisco) [reply tests/common/snappi_tests/snappi_fixtures.py]: And after fixing the RP, we see full pass with test_multidut_snappi.py:

```
=========================================================================================================================== PASSES ===========================================================================================================================
______________________________________________________________________________________________________________ test_snappi[multidut_port_info0] ______________________________________________________________________________________________________________
______________________________________________________________________________________________________________ test_snappi[multidut_port_info1] ______________________________________________________________________________________________________________
----------------------------------------------------------------------------- generated xml file: /run_logs/ixia/fixturize-reboot/2024-10-30-02-19-21/tr_2024-10-30-02-19-21.xml -----------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
02:25:43 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/test_multidut_snappi.py::test_snappi[multidut_port_info0]
PASSED snappi_tests/test_multidut_snappi.py::test_snappi[multidut_port_info1]
ERROR snappi_tests/test_multidut_snappi.py::test_snappi[multidut_port_info0] - Failed: Processes "['analyze_logs--<MultiAsicSonicHost yy39top-lc4>']" failed with exit code "1"
===================================================================================================== 2 passed, 5 warnings, 1 error in 379.58s (0:06:19) =====================================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$
```

File: tests/common/snappi_tests/qos_fixtures.py [hunk 2]
```diff
@@ -106,3 +111,55 @@ def lossy_prio_list(all_prio_list, lossless_prio_list):
...
+        for space in all_asic_list:
+            config = get_running_config(duthost, space)
+            if "PFC_WD" in config.keys():
+                all_configs.append(config['PFC_WD'])
+            else:
+                all_configs.append({})
+        return all_configs
+
+
+def reapply_pfcwd(duthost, pfcwd_config):
```
Comment (sdszhang) [inline tests/common/snappi_tests/qos_fixtures.py:134]: diffrence with existing function?

Comment (rraghav-cisco) [reply tests/common/snappi_tests/qos_fixtures.py:138]: @sdszhang , @amitpawar12 : Added a comment in the script.

**Comment by sdszhang** [comment]
@rraghav-cisco can you resolve the conflict. Thanks.

**Review by sdszhang — APPROVED** [review]
LGTM, ready to merge once conflict is resolved.

**Comment by yejianquan** [comment]
@rraghav-cisco , please raise a separated PR to 202405 branch

**Comment by rraghav-cisco** [comment]
> @rraghav-cisco PR conflicts with 202405 branch

@mssonicbld : https://github.com/sonic-net/sonic-mgmt/pull/15323

**Comment by rraghav-cisco** [comment]
> @rraghav-cisco , please raise a separated PR to 202405 branch

@yejianquan :https://github.com/sonic-net/sonic-mgmt/pull/15323

File: tests/snappi_tests/multidut/pfc/files/multidut_helper.py [hunk 3]
```diff
@@ -91,10 +91,6 @@ def run_pfc_test(api,
 
     pytest_assert(testbed_config is not None, 'Fail to get L2/3 testbed config')
 
-    for duthost, asic in dut_asics_to_be_configured:
-        stop_pfcwd(duthost, asic)
-        disable_packet_aging(duthost)
```
[the remark this answers is not in the document]
Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfc/files/multidut_helper.py:96]: @auspham , The same is being done as a fixture. Any test that needs to disable pfcwd should be calling that fixture now. This fixture takes care of re-applying the pfcwd configs if present in the DUT.
This avoids the config-reload done by the common setup/cleanup code, and saves 5 minutes per testcase. The fixture is: common/snappi_tests/qos_fixtures.py:disable_pfcwd()

[the remark this answers is not in the document]
Comment (sdszhang) [reply tests/snappi_tests/multidut/pfc/files/multidut_helper.py:96]: @rraghav-cisco can you open another PR to apply this fixture change for test_global_pause_with_snappi()?

[the remark this answers is not in the document]
Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfc/files/multidut_helper.py:96]: @sdszhang  , @auspham : Pls refer to: https://github.com/sonic-net/sonic-mgmt/pull/16613

**Pull request merged**